packages/core/src/theme.ts

```diff
@@ -9,26 +9,63 @@ import {
   ResultThemeTokens,
   SystemThemeTokens,
   InputThemeTokens,
+  NumberToken,
 } from "./themeTokens";
 import { componentList } from "./components/componentList";
-import { StyledProps } from "@kuma-ui/system";
+import { StyledProps, PseudoProps, ThemeSystemType } from "@kuma-ui/system";
 
-export type ThemeInput = InputThemeTokens & {
-  components?: {
-    [_ in keyof typeof componentList]?: {
-      baseStyle?: StyledProps;
-      variants?: { [key: string]: StyledProps };
-      defaultProps?: { variant?: string };
-    };
-  };
+type StyleProps = StyledProps & PseudoProps;
+
+type RawThemeComponent = {
+  baseStyle?: StyleProps;
+  variants?: { [key: string]: StyleProps };
+  defaultProps?: { variant?: string } & Record<string, unknown>;
+};
+
+type RawThemeComponents = {
+  [_ in keyof typeof componentList]?: RawThemeComponent;
+};
+
+export type RawThemeInput = InputThemeTokens & {
+  components?: RawThemeComponents;
+};
+
+type ThemeComponent<T> = {
+  [K in keyof T]?: K extends "baseStyle"
+    ? StyleProps
+    : K extends "variants"
+    ? {
+        [_ in keyof T[K] extends string ? keyof T[K] : never]: StyleProps;
```

Comment:
手元だとこう削っても型検査動いたんですがどうでしょう？
```suggestion
        [_ in keyof T[K]]: StyleProps;
```